### PR TITLE
Fixes for unit test fails

### PR DESCRIPTION
**Pull request by sumesh13**
Fixes include
1. Using LLVM APIs to compile to .o 
2. Load/Unload/Compile test case objects 
3. Proper insertion point for alloca/non-alloca instructions

### CODE REVIEW

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc [hunk 1]
```diff
@@ -250,65 +254,25 @@ std::vector<uint8> EmitModuleToHsaco(Module* module, llvm::TargetMachine* target
...
-    llvm_ir::AsStringRef("-o"),
-    llvm_ir::AsStringRef("isabin_path"),
-  };
-  if (inject_ir) {
-    llc_args[5] = llvm_ir::AsStringRef(inject_ir_path.c_str());
-  } else {
-    llc_args[5] = llvm_ir::AsStringRef(ir_path.c_str());
-  }
-  llc_args[7] = llvm_ir::AsStringRef(isabin_path.c_str());
+    // The extension is stripped by IrDumpingPassManager, so we need to
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc]: @sumesh13 indentation here is incorrect. could you shift 2 bytes to the left?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 2]
```diff
@@ -993,14 +993,15 @@ Status IrEmitterUnnested::EmitColumnReduction(
...
-
+        b_.SetInsertPoint(&entry_bb, entry_bb.begin());
         llvm::Value* partial_reduction_result_address =
             Alloca(element_ir_type, /*ArraySize=*/nullptr,
                    "partial_reduction_result." +
                        llvm::Twine(i * kTileWidth + x_offset));
+        b_.SetInsertPoint(old_insert_bb, old_insert_ip);
         TF_ASSIGN_OR_RETURN(llvm::Value* const init_ir_value,
                             init_value_gens[i](IrArray::Index(index_ty)));
         b_.SetInsertPoint(old_insert_bb, old_insert_ip);
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: would this line be obsolete?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 3]
```diff
@@ -993,14 +993,15 @@ Status IrEmitterUnnested::EmitColumnReduction(
 
     for (int i = 0; i != num_reduces; ++i) {
       for (int x_offset = 0; x_offset < kTileWidth; ++x_offset) {
+
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: could you remove this empty line?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 4]
```diff
@@ -1337,19 +1338,21 @@ Status IrEmitterUnnested::EmitRowReduction(
...
+      
+      b_.SetInsertPoint(old_insert_bb, old_insert_ip);
+
       TF_ASSIGN_OR_RETURN(llvm::Value* const init_ir_value,
                           init_value_gens[i](IrArray::Index(index_ty)));
-      b_.SetInsertPoint(old_insert_bb, old_insert_ip);
       Store(init_ir_value, partial_reduction_result_address);
       partial_reduction_result_addresses.push_back(
           partial_reduction_result_address);
+
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: please remove this line

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc [hunk 5]
```diff
@@ -56,6 +56,7 @@ limitations under the License.
 #include "llvm/Transforms/IPO/AlwaysInliner.h"
 #include "llvm/Transforms/IPO/PassManagerBuilder.h"
 #include "llvm/Transforms/IPO/Internalize.h"
+//#include "lld/common/Driver.h"
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc]: please remove this line and submit codes for lld in later PRs

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc [hunk 6]
```diff
@@ -222,6 +223,9 @@ std::vector<uint8> EmitModuleToHsaco(Module* module, llvm::TargetMachine* target
   std::string isabin_filename = tensorflow::strings::StrCat(module->getModuleIdentifier(), ".o");
   std::string isabin_path = tensorflow::io::JoinPath(tempdir_name, isabin_filename);
 
+  std::string isaasm_filename = tensorflow::strings::StrCat(module->getModuleIdentifier(), ".s");
+  std::string isaasm_path = tensorflow::io::JoinPath(tempdir_name, isaasm_filename);
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc]: please remove unused codes

**Comment by whchung** [comment]
@sumesh13 most of the changes i proposed are cosmetic. please help address them and I'd love to take it.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc [hunk 7]
```diff
@@ -322,15 +281,15 @@ std::vector<uint8> EmitModuleToHsaco(Module* module, llvm::TargetMachine* target
     llvm_ir::AsStringRef("isabin_path"),
     llvm_ir::AsStringRef("-o"),
     llvm_ir::AsStringRef("hsaco_path"),
-  };
+ };
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/amdgpu_backend_lib.cc]: nit pick: indent incorrect

**Comment by sunway513** [comment]
@whchung  Do I need to cherry-pick this PR to r1.12-rocm branch if XLA won't be abled there?

**Comment by whchung** [comment]
@sunway513 In case we want to enable XLA in 1.12 release, please help cherry-pick this into r1.12-rocm branch. Otherwise it's not needed.

**Comment by sunway513** [comment]
I don't think we'll enable XLA in 1.12, so I'll ignore it for now.

**Pull request merged**